cpp/include/graph.hpp

```diff
@@ -0,0 +1,213 @@
+   * @Param  edge_data             This array size E (number of edges) contains the weight for each edge.  This array can be null
+   *                               in which case the graph is considered unweighted.
+   * @Param  number_of_vertices    The number of vertices in the graph
+   * @Param  number_of_edges       The number of edges in the graph
+   */
+  GraphCOO(VT const *src_indices_, VT const *dst_indices_, WT const *edge_data_,
+           size_t number_of_vertices_, size_t number_of_edges_):
+    GraphBase<WT>(edge_data_, number_of_vertices_, number_of_edges_),
+    src_indices(src_indices_),
+    dst_indices(dst_indices_)
+  {}
+};
+
+/**
+ * @Synopsis    Base class for graph stored in CSR (Compressed Sparse Row) format.
+ *
+ * @tparam VT   Type of vertex (defaults to int)
+ * @tparam WT   Type of weight (defaults to float)
+ */
+template <typename VT = int, typename WT = float>
+class GraphCSRBase: public GraphBase<WT> {
+public:
+  VT const *offsets;       ///< CSR offsets
+  VT const *indices;       ///< CSR indices
+
+  /**
+   * @Synopsis    Fill the identifiers array with the vertex identifiers.
+   *
+   * @param[out]    identifier      Pointer to device memory to store the vertex identifiers
+   */
+  void get_vertex_identifiers(VT *identifiers) const;
+  
+  /**
+   * @Synopsis    Fill the identifiers in the array with the source vertex identifiers
+   *
+   * @param[out]    src_indices      Pointer to device memory to store the source vertex identifiers
+   */
+  void get_source_indices(VT *src_indices) const;
+
+  /**
+   * @Synopsis   Wrap existing arrays representing adjacency lists in a Graph.
+   *             GraphCSR does not own the memory used to represent this graph. This
+   *             function does not allocate memory.
+   *
+   * @Param  offsets               This array of size V+1 (V is number of vertices) contains the offset of adjacency lists of every vertex.
+   *                               Offsets must be in the range [0, E] (number of edges).
+   * @Param  indices               This array of size E contains the index of the destination for each edge.
+   *                               Indices must be in the range [0, V-1].
+   * @Param  edge_data             This array of size E (number of edges) contains the weight for each edge.  This
+   *                               array can be null in which case the graph is considered unweighted.
+   * @Param  number_of_vertices    The number of vertices in the graph
+   * @Param  number_of_edges       The number of edges in the graph
+   */
+  GraphCSRBase(VT const *offsets_, VT const *indices_, WT const *edge_data_,
+               size_t number_of_vertices_, size_t number_of_edges_):
+    GraphBase<WT>(edge_data_, number_of_vertices_, number_of_edges_),
+    offsets(offsets_),
+    indices(indices_)
+  {}
+};
+
+/**
+ * @Synopsis    A graph stored in CSR (Compressed Sparse Row) format.
```

Comment:
What's the difference between GraphCSR and GraphCSRBase? Shouldn't we document those?

Reply:
Renamed things to try and clarify.